### PR TITLE
Add modularity for easier widget creation

### PR DESCRIPTION
**Pull request by ltsSmitty**
Hi @KatieZeldaKat , I decided to take a swing at making some contributions. I created a structural system for making it easier to add new statistics to track, to hopefully lower the bar even further to contributing. 

Key additions:
* created a Statistics object with well-defined arguments to make it easy to add a new stat for tracking. Extracts the widget creation and management process.
* create widgets for tracking crashed vehicles and rides built
* CONTRIBUTION has a well documented example of how to create additional statistics.

Notes:
* the plugin api appears to have broken for `ridecreate` events. It is always firing at the right time, but it is also firing 1-3 extra times when creating new rides. I've added a fix that will remove these erroneous numbers, but as a result it may under-count by a small fraction. This will likely need upstream changes to the codebase to solve longterm.
* I just copied exactly your plugin widget layout and design. I could see someone wanting to customize their widget to display something else, but that's currently unsupported. This structure could be extended to support custom widget designs, but I wanted to run this by you before putting in that work (and since I don't know if anyone would care).

Happy to chat on discord by dm or in opencrt @ltsSmitty! Let me know if you have any thoughts.

### CODE REVIEW

**Comment by ltsSmitty** [comment]
Here are some visuals of the plugin with my two new widgets.

image

**Comment by KatieZeldaKat** [comment]
I appreciate the time you took to make these changes; there's a lot here that really reduces the work needed to add a new statistic! Probably the biggest thing your changes made me realize is I don't really have a well defined code style guide (before I was generally using a combination of style guides I had combined from memory of `GDScript` and `TypeScript`). I was tempted to create one in lieu of this, but already there's a lot to evaluate in terms of this request, so I will leave it as is and accept the modifications you made in that regard. Perhaps I'll eventually make another PR that will specify guidelines in a dedicated document combined with CI to automatically catch style mismatches.

I've been looking over the code changes for quite a few hours now, and while generally I agree with the implementation, there are various changes I will request. Most of them I'll try to separate into comments with line references so it's easier to fix one at a time. There is one overarching change I want to bring up, however:

"Rides Built" and "Vehicles Crashed" - I appreciate seeing these statistics implemented, as it does help to get a sense of how the general statistic tracking system works. However, I prefer these being in their own pull requests. Especially in the case of "Rides Built", having it delegated to another PR means that it can potentially be put on hold while bugs are fixed in OpenRCT2 without putting every single other change on hold as well.

Anyway, I'm going to try and start putting together my notes in a coherent way so we can discuss and get this new system fully down pat. Thanks again :)

File: src/startup.ts
Comment (KatieZeldaKat) [inline src/startup.ts:32]: If you rename `sc` to `statController`, then having the `.add()` calls on other lines would look quite clean and allow adding more stats easier:

```ts
statController
  .add(timeSpentStat)
  .add(ridesBuiltStat)
  .add(vehiclesCrashedStat);
```

File: src/statistics/timeSpent.ts
Comment (KatieZeldaKat) [inline src/statistics/timeSpent.ts:44]: Given the changes I proposed for `Statistic`, the exact code may need to change, but I will still propose legacy support for time tracking, since the keys are different.

```suggestion
  const statistic = new Statistic({
    key,
    title,
    statStore,
    resetValue: 0,
    formatDisplay,
    accumulator: accumulateSeconds,
  });
  
  // LEGACY SUPPORT FOR TRACKING TIME IN-GAME
  const legacyGameTimeKey = "openrct2-statistics.time.gameTime";
  if (context.sharedStorage.has(legacyGameTimeKey)) {
    // To prevent a conflict when importing the shared storage, wait a second beforehand
    context.setTimeout(() => {
      let legacyTime = context.sharedStorage.get(legacyGameTimeKey) as number;
      if (legacyTime >= 0) {
        statistic.gameStatStore.set(legacyTime + 1);

        // Prevent legacy value from being imported more than once
        context.sharedStorage.set(legacyGameTimeKey, -1);
      }
    }, 1000);
  }

  // LEGACY SUPPORT FOR TRACKING TIME IN PARKS
  context.subscribe("map.changed", () => {
    const legacyParkTimeKey = "openrct2-statistics.time.parkTime";
    if (context.mode == "normal" && context.getParkStorage().has(legacyParkTimeKey)) {
      // To prevent a conflict when importing the park storage, wait a second beforehand
      context.setTimeout(() => {
      let legacyTime = context.getParkStorage().get(legacyParkTimeKey) as number;
        if (legacyTime >= 0) {
          statistic.parkStatStore.set(legacyTime + 1);

          // Prevent legacy value from being imported more than once
          context.getParkStorage().set(legacyParkTimeKey, -1);
        }
      }, 1000);
    }
  });

  return statistic;
```

**Review by KatieZeldaKat — CHANGES_REQUESTED** [review]
I have quite a few comments, but a lot of them are minor. The biggest thing is that `Statistic` should be modified so that it does not rely on a separate `Store` to update both the game and park values.

File: src/objects/Statistic.ts
Comment (ltsSmitty) [reply src/objects/Statistic.ts:3]: thanks for pointing that out. probably done by Copilot and I never looked at it again.

File: src/objects/Statistic.ts
Comment (ltsSmitty) [reply src/objects/Statistic.ts:120]: just kidding, you're right

File: src/statistics/timeSpent.ts
Comment (ltsSmitty) [reply src/statistics/timeSpent.ts:44]: in line 56 and 72, this sets the -1 - wouldn't it be better to set to `undefined` so that after the initial load it'll never find the key again?

File: src/window/createStatWidget.ts
Comment (ltsSmitty) [reply src/window/createStatWidget.ts:32]: I went back and forth on this. Making it optional like a possible way to decrease the initial lift of creating a new stat. hence why I added the JSON.stringify fallback. I think this is easy enough that instead we can add a few examples of what this could look like (which are actually already done pretty well with the time example and the two that i created) and trust the user to get it.

File: src/window/createStatWidget.ts
Comment (ltsSmitty) [reply src/window/createStatWidget.ts:53]: yes! when basssiiie showed it to me I was so glad to see it.

File: src/window/playTimeWidget.ts
Comment (ltsSmitty) [reply src/window/playTimeWidget.ts:1]: correct, this was my first attempt to convert your time stat into this new format. will delete.

File: src/objects/Statistic.ts
Comment (ltsSmitty) [reply src/objects/Statistic.ts:15]: i very much see what you're saying here. let me try reimplementing this and see what the ergonomics are like.

**Comment by ltsSmitty** [comment]
I've gone ahead and made all the requested changes, waiting on your feedback on if we should set the values for legacy items to -1 or to undefined. woooooo

File: src/statistics/timeSpent.ts
Comment (KatieZeldaKat) [reply src/statistics/timeSpent.ts]: To be honest, I was under the assumption that after you set a key, you wouldn't be able to remove it (since there's no method for that, just to `set`). If that's the case, then yeah, `undefined` would make sense, and there wouldn't need to be checks for `legacyTime >= 0` anymore.

File: src/statistics/timeSpent.ts
Comment (KatieZeldaKat) [inline src/statistics/timeSpent.ts:47]: When setting to `undefined`, then the following resolves to `false`:

```ts
context.sharedStorage.has(legacyGameTimeKey)
```

The key is deleted in the `plugin.store.json` file too. I'd imagine it's similar in a `.park` file, though I wouldn't be sure how to check since it's not a text file.

The comment can likely be removed and the code should be updated to the simpler expression. Same for legacy support for parks.

File: src/statistics/timeSpent.ts
Comment (KatieZeldaKat) [inline src/statistics/timeSpent.ts:52]: Not needed anymore since setting to `undefined` will delete the time value altogether. Same for legacy support for parks.

File: src/startup.ts
Comment (KatieZeldaKat) [inline src/startup.ts:18]: Missed this last time. This is the legacy time instance. It shouldn't be imported nor initialized (honestly, the legacy statistic and window can probably be deleted, as if they need to be referenced, they are [in the v1.0.0 tag](https://github.com/KatieZeldaKat/openrct2-statistics/tree/v1.0.0)).

File: src/window/window.ts
Comment (KatieZeldaKat) [inline src/window/window.ts:20]: I'd remove these comments.

File: README.md
Comment (KatieZeldaKat) [inline README.md:7]: All changes in this file should be reverted and only applied to the other PRs that would add the functionality.

**Review by KatieZeldaKat — CHANGES_REQUESTED** [review]
Changes this time around are more minor, apart from `time.ininitialize()` still being called. Feel free to update `CONTRIBUTING.md`. That will definitely need updating, and I don't believe any significant changes will happen in the code that warrant it being put on the backburner.

**Comment by ltsSmitty** [comment]
Did these revisions. I'm pleased with how clean the code is at this point. 

While I was at it, I cleaned up the windowPause code to use compute() instead of new dedicated stores. I think you'll find it much cleaner.

**Comment by ltsSmitty** [comment]
Finally, I updated the contribution guide to accomodate the changes. I'd like to add a few more examples, but I want to use existing code from the codebase, so that'll have to wait for a further PR to add additional stats, so that the examples come from code that can be actually referenced.

File: docs/CONTRIBUTING.md
Comment (KatieZeldaKat) [inline docs/CONTRIBUTING.md:17]: An instance of `TimeSpendStat` that wasn't caught. Which leads into one point I think would help maintaining this `CONTRIBUTING.md` file: There should probably be less code blocks.

Any time code is copy-pasted here, there's the potential this document will become outdated as changes are made. Perhaps having more instances of something along the lines of "See `subscribeToTimePassing` in [`timeSpent.ts`](../src/statistics/timeSpent.ts)" would help to reduce the amount of potential changes.

If simple enough, examples can perhaps be here (honestly, this example of `TimeSpendStat` was less a critisim of using direct copy-paste here and more an example of why this could cause problems in the future; it illistrates this point well), but most examples, especially longer function implementations, might be best served in their original files.

**Review by KatieZeldaKat — CHANGES_REQUESTED** [review]
Just a few more things. Went through each file individually this time around and I don't think there'd be anything else code-wise. Given my requests on the `CONTRIBUTING.md` file though, I wouldn't be surprised if we got a few more passes on that in particular.

File: src/objects/Statistic.ts
Comment (ltsSmitty) [reply src/objects/Statistic.ts:51]: I prefer to use a props object when writing new code, especially when I'm unsure what the final shape of a function and its args is going to look like. For me this reduces future errors/debugging by not relying on the positioning of an argument and instead referencing it by name. (At this point in the development of this codebase, I no longer have the same questions of structure and would be willing to remove it, but I almost always start development at this level and don't dislike its appearance enough to go back and change.)

I think stylistically it also matches what's being done at the display layer with FlexUI expecting args as props. This is not a hill I'll die on, just a personal preference from my experiece & writing React. I'm willing to change this and `createStatWidget` if it's important to you.

File: src/window/createStatWidget.ts
Comment (KatieZeldaKat) [inline src/window/createStatWidget.ts:27]: I feel like renaming this to `formatDisplay` for consistency would make sense. It's also more descriptive.

File: src/objects/Statistic.ts
Comment (KatieZeldaKat) [reply src/objects/Statistic.ts:51]: I've thought about it for a little bit, and I think that we can remove `props` from `Statistic`, but you're right that the FlexUI framework does use objects for parameters, so it makes sense that `createStatWidget` would do similarly.

File: src/window/createStatWidget.ts
Comment (KatieZeldaKat) [inline src/window/createStatWidget.ts:16]: Let's rename `props` to `params` to match the definitions in FlexUI.

File: docs/CONTRIBUTING.md
Comment (KatieZeldaKat) [inline docs/CONTRIBUTING.md:92]: I believe this should be referring to Step 6?

File: src/objects/Statistic.ts
Comment (KatieZeldaKat) [inline src/objects/Statistic.ts:110]: After some testing, `isPaused` was not being initialized correctly when booting up the game after having stats paused. We could change its definition to `isPaused = areStatisticsPaused.get()`, but at that point, why even have a separate variable? Just delete `isPaused` from the file and change this line to this:

```suggestion
    if (areStatisticsPaused.get()) {
```

**Review by KatieZeldaKat — CHANGES_REQUESTED** [review]
Was about to approve, but found a bug when testing I didn't notice before.

**Comment by ltsSmitty** [comment]
I see, it looks like the pause widget is initializing after the Statistics ones, causing the initial value for the statistic store to be undefined. I approve of your solution and made that change.

While I was at it, I thought it made sense to swap the pause widget from using a button to using a toggle. using compute() makes tracking the state of the toggle press simple and I think the UI of a toggle matches the functionality better than a button.

**Review by KatieZeldaKat — CHANGES_REQUESTED** [review]
I don't see the need to change the pause button to a toggle. The sprite icon changing was meant to be the way to signify the status being different, much like when opening or closing the park or a ride.

The back of the icon now lights up bright at all times when stats are paused, which I don't really think looks very good, and it isn't what a flag to open or close the park or a ride looks like when pressed.

**Comment by ltsSmitty** [comment]
I see. I compare it to the pause toggle in the top bar, which has this new behavior of background changing to display the pause, but I see your rationale. I'll revert.

**Comment by ltsSmitty** [comment]
reverted, and everything addressed as far as i can tell. thanks for your involvement and diligence on this.

**Review by KatieZeldaKat — APPROVED** [review]
Looks good, doesn't seem like there are any lingering issues or bugs. This should really make it easier to create new statistics to track in the future, great work.

I'll draft up a new release so people will be using the updated save keys, even if their experience with the plugin won't change from before.

**Pull request merged**
